presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -717,13 +732,15 @@ private ConnectorTableMetadata doGetTableMetadata(ConnectorSession session, Sche
 
     @SuppressWarnings("TryWithIdenticalCatches")
     @Override
-    public Map<SchemaTableName, List<ColumnMetadata>> listTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    public Stream<ListTableColumnsResult> listTableColumnsStream(ConnectorSession session, SchemaTablePrefix prefix)
     {
         requireNonNull(prefix, "prefix is null");
-        ImmutableMap.Builder<SchemaTableName, List<ColumnMetadata>> columns = ImmutableMap.builder();
-        for (SchemaTableName tableName : listTables(session, prefix)) {
+        return listTables(session, prefix).stream().map(tableName -> {
             try {
-                columns.put(tableName, getTableMetadata(session, tableName).getColumns());
+                if (redirectTable(session, tableName).isPresent()) {
```

Comment:
Add a check in `getTableHandle` and `getTableHandleForStatisticsCollection` like
```
verify(redirectTable(...).isEmpty(), "Table access should have been redirected for %s", tableName);
```

This will make sure (or improve probability that) if we miss to add `metadata.redirectTable` somewhere (in existing or new code), this will not lead to incorrect results or data corruption.

Reply:
Good idea. Added in them as well as some other methods.